### PR TITLE
docs: API review checklist

### PR DESCRIPTION
**Pull request by markdroth**
Commit Message: docs: API review checklist
Additional Description: Checklist to be used by API reviewers and by those who file PRs to make API changes.
Risk Level: Low
Testing: N/A
Docs Changes: Included in PR
Release Notes: N/A
Platform Specific Features: N/A

CC @envoyproxy/api-shepherds

### CODE REVIEW

File: api/review_checklist.md
Comment (alyssawilk) [inline api/review_checklist.md:5]: mind linking to this from CONTRIBUTING.md too?

"If your PR includes changes to api/... please read "

File: api/review_checklist.md
Comment (alyssawilk) [inline api/review_checklist.md:10]: I think much of this overlaps in spirit with the GOVERNANCE runtime guard section.  I know the API stuff is different since it affects more than just envoy, but it'd still be good to address the overlap, WDYT?

File: api/review_checklist.md
Comment (alyssawilk) [inline api/review_checklist.md:40]: I wonder if we can enforce any of these with fix_format / proto scripts.  If we had a required required/not_required annotation, or required "unbounded" annotations, so folks had to explicitly annotate, it'd be annoying but might catch more.
I'd file this under "food for thought" where you and the other shephards can decide if it's worth it after seeing what slips through review over time :-)

File: api/review_checklist.md
Comment (alyssawilk) [inline api/review_checklist.md:53]: Haven't we traditionally added a new field and deprecated old ones in the same PR?  I guess better project party is part of pulling the APIs out here.  I think it'd be helpful if before we codify this rule, we have some process around adding a field, making sure the migrations are done in both projects, being able to determine when they're done and having issues filed to do the deprecation.

**Review by alyssawilk — COMMENTED** [review]
Cool, so many good review points in this doc - it's great to have them written down!

File: api/review_checklist.md
Comment (htuch) [reply api/review_checklist.md:10]: Good point, I think we should port the relevant parts of GOVERNANCE.md across to an API GOVERNANCE.md; it might be copy+paste.

We are working towards a multi-stakeholder governance for the APIs (under CNCF, see https://github.com/cncf/udpa/blob/master/README.md), so we should evolve the API docs with that in mind.

This might seem a minor thing, but I think this decoupling is critical to ensuring that other adopters of xDS are comfortable and don't feel locked into Envoy project governance. We will do this while retaining the agree upon foundations from GOVERNANCE.md.

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:1]: What about adding a checkbox to https://github.com/envoyproxy/envoy/blob/master/PULL_REQUEST_TEMPLATE.md indicating that (for API changes) that they've understood this policy?

File: api/review_checklist.md
Comment (mattklein123) [inline api/review_checklist.md:36]: Should we just merge this into STYLE.md? Seems like we should have all of this in one place?

Also, something that came up recently, can you add something about making sure to use `bytes` vs. `string` when the data is actually raw bytes? We have had multiple problems with this.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this is great. One small comment from me and I think we can ship and iterate once we fold in all the other comments.

/wait

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:1]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:5]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:10]: Should we have a governance file in cncf/xds and just point at that from api/GOVERNANCE.md in this repo, to make it clear that it's a separate thing?  Actually, for that matter, should we do the same thing with files like this one and api/STYLE.md as well?

In any case, I agree that we should document the governance of the API and get all of these docs properly aligned, but I think a lot of this is outside the scope of this PR.  Can we follow up on that separately?

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md]: I moved this bullet to STYLE.md and added another one there about `bytes` vs. `string`.

The first two bullets in this section actually seem less about style and more about extensibility, so I kept them in this file but moved them down to the "Extensibility" section.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:47]: I think we can play this by ear.  Right now, there are only 2 clients, and the API shepherds all talk to each other, so it's fairly easy to coordinate this without much red tape.  Over time, if this becomes more of a problem, we can add more process as appropriate.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:34]: I'll let @htuch think about this one.

File: api/review_checklist.md
Comment (alyssawilk) [reply api/review_checklist.md:47]: Can we redact this for now or add an "or get sign off from each project stakeholder for immediate deprecation"?  Given n=2 I think that's less onerous and as we add more stakeholders I think we'd have to move towards the deprecation dance anyway.

File: api/review_checklist.md
Comment (alyssawilk) [reply api/review_checklist.md:10]: SGTM

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:47]: I've added a sentence indicating to ask the API shepherds if you're not sure about the status of a feature in the major known clients.

**Review by mattklein123 — APPROVED** [review]
Thanks LGTM to ship and iterate. Will defer to others on further review.

File: api/review_checklist.md
Comment (alyssawilk) [reply api/review_checklist.md:47]: So I'm on board with playing this by ear, but the documentation this still disallows changes like I mentioned. I want us to be explicit here.  I don't want to pedantic, but  I really want to avoid having codified rules we encourage people in the know to ignore.  If we're Ok playing it by ear, let's remove the restriction that you can't deprecate a field until the replacement is ready in all consuming stacks.  If we're not OK doing this with shephard approval, we should make it clear it is now disallowed to add A and deprecate B in the same PR.

File: api/review_checklist.md
Comment (htuch) [reply api/review_checklist.md:10]: SG, @markdroth can you file an issue on cncf/xds around this?

File: api/review_checklist.md
Comment (htuch) [reply api/review_checklist.md:34]: This could be done. I think if I had my wishtlist of tooling, we would have something that during code review annotates the PR when a new field is introduced and is missing constraints to verify. 

@markdroth can you be more explicit and say something like "Is the field mandatory? If so, are protocgen-validate constraints defined to enforce this?".

File: PULL_REQUEST_TEMPLATE.md
Comment (htuch) [inline PULL_REQUEST_TEMPLATE.md:25]: Can you GitHub Markdown link "API considerations" to `api/review_checklist.md`? Most folks will not read `CONTRIBUTING.md` line-by-line, so this is another opportunity to remind.

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:69]: One thing that I'd like to do a better job of is placing protos that might eventually become shared in shared locations to begin with. We've seen historically situations where we have had to copy+paste protos to common directories or refer to protos in strange locations (e.g. some rando filter referring to the access log matcher in the access log tree). This advice also includes thinking carefully about whether some sub-messages might be one day useful elsewhere. 

It's hard to be super prescriptive, but good to be thoughtful about. Basically, without major versioning as a device, refactors are messy.

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:82]: Another bullet point "If this change involves extension configuration, how will it interact with ECDS?" CC @kyessenov

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:94]: Can you link to https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md as an `e.g. for Envoy`? As per my earlier comment, we can't make this policy the API policy, but providing an example is probably good.

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:104]: I think config size scalability is also a huge deal. We should make sure we're careful not to add things that might blow out config size, e.g. some huge proto that is mandatory and attached to every single route entry, or things that are expensive during config ingestion, such as JSON round-trips.

File: api/review_checklist.md
Comment (htuch) [inline api/review_checklist.md:54]: Should we mention `TypedExtensionConfig` somewhere?

**Review by htuch — CHANGES_REQUESTED** [review]
This is awesome! I've left a few comments, but what is there looks great.
/wait

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:49]: Ah, okay, sorry -- I somehow didn't grok that it was the ability to explicitly sign off on exceptions that you wanted to call out.  I've added a clause about that.

File: PULL_REQUEST_TEMPLATE.md
Comment (markdroth) [reply PULL_REQUEST_TEMPLATE.md]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:10]: https://github.com/cncf/xds/issues/1

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:36]: I added a separate bullet about having PGV rules, since that basically applies to all of the bullets in this section.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:57]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:72]: Added a bullet about this.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:90]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md]: Done.

File: api/review_checklist.md
Comment (markdroth) [reply api/review_checklist.md:116]: Added a couple of bullets about this.

**Review by alyssawilk — DISMISSED** [review]
Awesome, thanks for that last tweak!

**Comment by alyssawilk** [comment]
rather LGTM for content - looks like CI needs a fix :-)

**Review by htuch — APPROVED** [review]
LGTM once build is fixed.

**Review by alyssawilk — APPROVED** [review]
LGTM!

Though it looked fine to me with the extra whitespace too, haha.

**Comment by markdroth** [comment]
I'm not sure what's causing the remaining CI failure.  Let me try merging main and see if that fixes it.

**Comment by markdroth** [comment]
The CI failures look unrelated to this PR.  Can someone please advise as to what's going on here?

**Comment by alyssawilk** [comment]
cc @asraa there's a known download issue now correct?  Any workarounds?

**Comment by asraa** [comment]
There's some discussion about dependency errors in CI on slack, but this one looks like it's a network error trying to install packages. Brought that on the slack thread :/
```error: failed to solve: rpc error: code = Unknown desc = executor failed running [/dev/.buildkit_qemu_emulator /bin/sh -c echo "d6c40440609a23483f12eb6295b5191e94baf08298a856bab6e15b10c3b82891  /tmp/su-exec.c" > /tmp/checksum     && curl -o /tmp/su-exec.c https://raw.githubusercontent.com/ncopa/su-exec/212b75144bbc06722fbd7661f651390dc47a43d1/su-exec.c     && sha256sum -c /tmp/checksum     && gcc -Wall /tmp/su-exec.c -o/usr/local/bin/su-exec     && chown root:root /usr/local/bin/su-exec     && chmod 0755 /usr/local/bin/su-exec]: exit code: 1
```

Re-running, I'll monitor this PR

**Pull request merged**
